lib/wallaroo/core/sink/connector_sink/connector_sink.pony

```diff
@@ -363,8 +576,130 @@ actor ConnectorSink is Sink
     end
     match barrier_token
     | let sbt: CheckpointBarrierToken =>
-      checkpoint_state(sbt.id)
+      if not _notify.twopc_intro_done then
+        // This sink applies Pony runtime backpressure when disconnected,
+        // so it's quite unlikely that we will get here: sending
+        // messages to us will block the senders.  However, the nature
+        // of backpressure may change over time as the runtime's
+        // backpressure system changes.
+        ifdef "checkpoint_trace" then
+          @printf[I32]("2PC: preemptive abort: connector sink not fully connected\n".cstring())
+        end
+        _abort_decision("connector sink not fully connected",
+          _twopc_txn_id, _twopc_barrier_token)
+
+        _twopc_txn_id = "preemptive txn abort"
+        _twopc_barrier_token = sbt
+        return
+      end
+
+      if _twopc_state is cp.TwoPCFsmStart then
+        let txn_id = _make_txn_id_string(sbt.id)
+        _twopc_txn_id = txn_id
+        // Correctness of the barrier protocol requires us to write
+        // checkpoint state here, not in barrier_fully_acked()
+        checkpoint_state(sbt.id)
+
+        if (_twopc_current_offset > 0) and
+           (_twopc_current_offset == _twopc_last_offset)
+        then
+          // If no data has been processed by the sink since the last
+          // checkpoint, then don't bother with 2PC.
+
+          ifdef "checkpoint_trace" then
+            @printf[I32]("2PC: no data written during this checkpoint interval, skipping 2PC round\n".cstring())
+          end
+          _barrier_coordinator.ack_barrier(this, sbt)
+          _twopc_txn_id = ""
+          _twopc_phase1_commit = true
+          _twopc_state = cp.TwoPCFsm2Commit
+          return
+        end
+
+        let where_list: cp.WhereList =
+          [(1, _twopc_last_offset.u64(), _twopc_current_offset.u64())]
+        let b = cp.TwoPCEncode.phase1(txn_id, where_list)
+        try
+          let msg = cp.MessageMsg(0, cp.Ephemeral(), 0, 0, None, [b])?
+           _notify.send_msg(this, msg)
+         else
+          Fail()
+        end
+        ifdef "checkpoint_trace" then
+          @printf[I32]("2PC: sent phase 1 for txn_id %s\n".cstring(), txn_id.cstring())
+        end
+
+        _twopc_state = cp.TwoPCFsm1Precommit
+        _twopc_txn_id = txn_id
+        _twopc_barrier_token = sbt
+      else
+        @printf[I32]("2PC: ERROR: _twopc_state = %d\n".cstring(), _twopc_state())
+        Fail()
+      end
+    end
+
+  be barrier_fully_acked(token: BarrierToken) =>
+    """
+    The BarrierInitiator has determined that the barrier protocol
+    for this autoscale/checkpoint/rollback/whatever event has
+    completed successfully, including all other 2PC protocol rounds
```

Comment:
This means the 1st 2PC protocol round has completed everywhere, but since this message is effectively the phase 2 "commit" message, I'm not sure it's right to say that all other 2PC protocol rounds were executed by other sinks.

Reply:
This docstring belonged to `barrier_fully_acked()` which has been removed.